### PR TITLE
partition distance

### LINKED ISSUE #1166: Partition distance metric
I think grouping and partitioning transformations should have an output metric of d_{P_M}, the "partition distance". It's not actually a metric; it outputs a tuple of length 3:

```
d_{P_M}(x, x') = (|d_M(x, x')|_0, |d_M(x, x')|_1, |d_M(x, x')|_\infty)
```
It measures how many partitions changed, total distance over all partitions, and max distance in any one partition

The stability map of partitioning or grouping is: `lambda d_in: (m, d_in, n)`
`m` and `n` come from:
* domain descriptors. Descriptors can tell you an individual only contributes to at most 3 partitions, or to each partition at most once, etc
* otherwise, establish by truncation
* otherwise, they are never worse than d_in

I have three kinds of transformations/measurements you might chain with afterwards:

# Vector Transformations
For Lp sensitivity, assuming homogeneous transformation `t` applied to each partition:
```python
def stability_map(d_in):
    l0, l1, linf = d_in
    approach1 = t.map(l1)
    approach2 = np.linalg.norm([t.map(linf)] * l0, ord=p)
    return min(approach1, approach2)
```

When `t` is a count, this results in
* an L2 sensitivity of `sqrt(l0 * linf^2)` when you have knowledge that an individual has limited contribution to each partition 
* an L2 sensitivity of `l1` when you don't: all edits could land in the same partition

# Parallel Composition
Assuming heterogeneous mechanisms `ms` applied to each partition:

```python
def stability_map(d_in):
    l0, l1, linf = d_in
    return l0 * max(m.map(linf) for m in ms)
```
You could make this tighter by taking the sum of the top `l0` privacy losses.

# Sample and Aggregate
```python
def stability_map(d_in):
    return d_in[0]
```
You're applying a black-box function to each partition, it doesn't care how much each partition can change

--------

For both vector queries (approach2) and parallel composition, you could make their maps tighter by limiting the number of changes to just l1, which can be ~a sqrt less than linf * l0. But the maps are not linear, they are affine. So in order to do this, you'd have to either:
take the max sensitivity over all distributions of edits over the partitions (computationally infeasible)
know that the map is an affine transformation and retrieve the slope/intercept from it

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #1166

### CODE REVIEW

File: rust/src/ffi/util.rs [hunk 1]
```diff
@@ -289,6 +289,8 @@ lazy_static! {
             // OptionDomain<AtomDomain<_>>::Carrier
             type_vec![[Vec Option], <bool, char, u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, usize, f32, f64, String, AnyObject>],
             type_vec![Vec, <(f32, f32), (f64, f64)>],
+            vec![t!((u32, u32, u32)), t!((u32, u64, u64)), t!((u32, i32, i32)), t!((u32, i64, i64))],
```
Comment (mccalluc) [inline rust/src/ffi/util.rs:292]: Might be obvious, but maybe add a comment explaining why larger or smaller integer types aren't needed?

**Review by mccalluc — COMMENTED** [review]
Is there any argument for using a struct instead of a tuple? Worry a little as tuples get a larger and the user needs to remember the meaning of each slot.

File: rust/src/data/ffi.rs [hunk 2]
```diff
@@ -181,14 +199,16 @@ pub extern "C" fn opendp_data__slice_as_object(
...
-        _ => dispatch!(
+        v => {
+            println!("{:?}", v);
+            dispatch!(
             raw_to_plain,
             [(
                 T,
                 [u8, u32, u64, u128, i8, i16, i32, i64, i128, usize, f32, f64, bool, AnyMeasurement, AnyQueryable]
             )],
             (raw)
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/data/ffi.rs:210]: Rustfmt is super conservative when it comes to changing code within macro invocations.

**Comment by mccalluc** [comment]
Thought this would be used for Tulap, but not needed in the end. Maybe useful for Polars? Low priority.

**Comment by Shoeboxam** [comment]
The main thing I think is that structs will be more difficult to carry around over ffi.

**Review by mccalluc — DISMISSED** [review]
Can't see the latest changes easily, but spot-checked where I had comments earlier, and it looks good. Approved, but if you'd like a re-review from the top, I could do that.

**Pull request merged**